From 4fd83a3b84b7565bb0a568b87a9aa5d9b2d2e67b Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Thu, 24 Feb 2011 13:42:34 -0700
Subject: [PATCH 012/177] - fixed bug in c2_cob_unlink() in error path
 (cursor_fini was missed); - added c2_cob_rename(). Use it in
 c2_md_store_rename(); - fixed md.ff fop format for rename. It missed target
 body.

---
 cob/cob.c         | 68 +++++++++++++++++++++++++++++++++++++++++++++++++++++--
 cob/cob.h         | 13 +++++++++++
 mdstore/md.c      |  4 ++--
 mdstore/md.ff     |  7 +++---
 mdstore/mdstore.c | 25 ++++++++++++++++----
 5 files changed, 106 insertions(+), 11 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index 4055f88..3d90538 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -807,11 +807,10 @@ int c2_cob_unlink(struct c2_cob        *cob,
 	                         &oikey, sizeof oikey, &nkey, sizeof nkey);
 
                 rc = c2_db_cursor_get(&cursor, &pair);
+                c2_db_cursor_fini(&cursor);
                 if (rc)
                         goto out;
 
-                c2_db_cursor_fini(&cursor);
-
                 /*
                  * Now let's move statdata itself using found namespace
                  * key and nsrec from cob. Note that nlink-- is realdy
@@ -874,6 +873,71 @@ out:
         return rc;
 }
 
+/**
+   Rename oldkey with passed newkey.
+   
+   cob    - stat data (zero name) cob;
+   srckey - source name;
+   tgtkey - target name;
+   tx     - transcation handle
+*/
+int c2_cob_rename(struct c2_cob        *cob, 
+                  struct c2_cob_nskey  *srckey,
+                  struct c2_cob_nskey  *tgtkey,
+                  struct c2_db_tx      *tx)
+{
+        struct c2_cob_nsrec nsrec;
+        struct c2_db_pair   pair;
+        struct c2_db_cursor cursor;
+        struct c2_cob_oikey oikey;
+        int                 rc;
+
+        rc = c2_db_cursor_init(&cursor, 
+                               &cob->co_dom->cd_namespace, tx);
+        if (rc)
+                return rc;
+
+        /* Let's check if tgt name already exists */
+        c2_db_pair_setup(&pair, &cob->co_dom->cd_namespace,
+	                 tgtkey, c2_cob_nskey_size(tgtkey),
+		         &nsrec, sizeof nsrec);
+        rc = c2_db_cursor_get(&cursor, &pair);
+        if (rc == 0) {
+                c2_db_cursor_fini(&cursor);
+                return -EEXIST;
+        }
+
+        /* Let's check if src name exists */
+        c2_db_pair_setup(&pair, &cob->co_dom->cd_namespace,
+	                 srckey, c2_cob_nskey_size(srckey),
+		         &nsrec, sizeof nsrec);
+
+        rc = c2_db_cursor_get(&cursor, &pair);
+        if (rc) {
+                c2_db_cursor_fini(&cursor);
+                return -ENOENT;
+        }
+
+        /* Update name in namespace */
+        c2_db_pair_setup(&pair, &cob->co_dom->cd_namespace,
+	                 tgtkey, c2_cob_nskey_size(tgtkey),
+		         &nsrec, sizeof nsrec);
+        rc = c2_db_cursor_set(&cursor, &pair);
+        c2_db_cursor_fini(&cursor);
+
+        /* Update object index */
+        oikey.cok_stobid = cob->co_stobid;
+        oikey.cok_linkno = nsrec.cnr_linkno;
+
+        c2_db_pair_setup(&pair, &cob->co_dom->cd_object_index,
+                         &oikey, sizeof oikey, &tgtkey, sizeof tgtkey);
+        rc = c2_table_update(tx, &pair);
+        c2_db_pair_release(&pair);
+        c2_db_pair_fini(&pair);
+
+        return rc;
+}
+
 /** @} end group cob */
 
 /*
diff --git a/cob/cob.h b/cob/cob.h
index dd35bd4..a30e031 100644
--- a/cob/cob.h
+++ b/cob/cob.h
@@ -286,6 +286,19 @@ int c2_cob_delete(struct c2_cob *cob,
                   struct c2_db_tx *tx);
 
 /**
+   Rename oldkey with passed newkey.
+   
+   cob    - stat data (zero name) cob;
+   srckey - source name;
+   tgtkey - target name;
+   tx     - transcation handle
+*/
+int c2_cob_rename(struct c2_cob        *cob, 
+                  struct c2_cob_nskey  *srckey,
+                  struct c2_cob_nskey  *tgtkey,
+                  struct c2_db_tx      *tx);
+                  
+/**
    Acquires an additional reference on the object.
 
    @see c2_cob_put()
diff --git a/mdstore/md.c b/mdstore/md.c
index 6beed11..249fc9b 100644
--- a/mdstore/md.c
+++ b/mdstore/md.c
@@ -53,8 +53,8 @@ static void c2_md_store_fop_free(struct c2_fop *fop)
                 break;
         case C2_FOP_RENAME:
                 rename = c2_fop_data(fop);
-                c2_free(rename->r_src.n_name);
-                c2_free(rename->r_tgt.n_name);
+                c2_free(rename->r_sname.n_name);
+                c2_free(rename->r_tname.n_name);
                 break;
         default:
                 break;
diff --git a/mdstore/md.ff b/mdstore/md.ff
index 6969443..4715779 100644
--- a/mdstore/md.ff
+++ b/mdstore/md.ff
@@ -43,9 +43,10 @@ DEF(c2_fop_unlink, RECORD,
     _(u_name, c2_fop_name));
 
 DEF(c2_fop_rename, RECORD,
-    _(r_body, c2_fop_body),
-    _(r_src, c2_fop_name),
-    _(r_tgt, c2_fop_name));
+    _(r_sbody, c2_fop_body),
+    _(r_tbody, c2_fop_body),
+    _(r_sname, c2_fop_name),
+    _(r_tname, c2_fop_name));
 
 DEF(c2_fop_open, RECORD,
     _(o_body, c2_fop_body));
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 4a07172..abe4f29 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -248,15 +248,32 @@ int c2_md_store_close(struct c2_md_store *md, struct c2_fop *fop,
 int c2_md_store_rename(struct c2_md_store *md, struct c2_fop *fop,
                        struct c2_fop_ctx *ctx)
 {
-        struct c2_fop_rename *rename;
-        int rc;
+        struct c2_fop_rename *rename = c2_fop_data(fop);
+        struct c2_fop_body   *sbody = &rename->r_sbody;
+        struct c2_fop_body   *tbody = &rename->r_tbody;
+        struct c2_cob_nskey  *srckey;
+        struct c2_cob_nskey  *tgtkey;
+        struct c2_stob_id     stobid;
+        struct c2_cob        *cob;
+        int                   rc;
         
         rc = c2_md_store_check_perm(md, fop, ctx);
         if (rc)
                 return rc;
 
-        rename = c2_fop_data(fop);
-        return 0;
+        fid2stobid(&stobid, &sbody->b_tfid);
+        make_nskey(&srckey, &sbody->b_pfid, &rename->r_sname);
+        make_nskey(&tgtkey, &tbody->b_pfid, &rename->r_tname);
+
+        rc = c2_cob_locate(&md->md_cob, &stobid, &cob, ctx->ft_tx);
+        if (rc)
+                return rc;
+                
+        rc = c2_cob_rename(cob, srckey, tgtkey, ctx->ft_tx);
+        c2_free(srckey);
+        c2_free(tgtkey);
+        
+        return rc;
 }
 
 int c2_md_store_setattr(struct c2_md_store *md, struct c2_fop *fop,
-- 
1.8.3.2

